### PR TITLE
Snippet instrumentation

### PR DESCRIPTION
**Pull request by digitarald**
@jrieken I am trying to extend @JacksonKearl's instrumentation to understand extension snippets. I got the property into the Snippets instance but my TS foo fails to access the `snippets` object on the `CompletionItem | SnippetCompletion`. Maybe I am holding it wrong.

Can you confirm if this approach is generally sane and how I solve the TS puzzle to access `.snippets`.

### CODE REVIEW

**Review by jrieken — CHANGES_REQUESTED** [review]
I think the approach can only work when you define the extension/source property on the completion model, e.g on the [`CompletionItem`](https://github.com/microsoft/vscode/blob/44ffeba23714e9333aa5e68aa53685faeb280bf5/src/vs/editor/common/languages.ts#L596)-type. With that each item would know its source. Today, that's done with the pointer to the [provider]([provider](https://github.com/microsoft/vscode/blob/98a51d5ef7194863c280562661b64775ca7fe246/src/vs/editor/contrib/suggest/browser/suggest.ts#L77)) of an item. That pointer doesn't work for snippets because there is only one snippet provider. That brings me to a potential alternative: Instead of a single snippet completion provider there could an instance per extension (that contributes snippets statically). That would slot nicely into the existing model but would be more work in snippet land...

**Comment by digitarald** [comment]
@jrieken I got your recommendation to track extensions in `CompletionItem` working. It seems lean enough, but what do you think?

**Review by jrieken — APPROVED** [review]
Small nit but the general direction is looking good.

**Review by raghavthind2005 — APPROVED** [review]

File: src/vs/editor/common/languages.ts [hunk 1]
```diff
@@ -673,6 +673,10 @@ export interface CompletionItem {
 	 * A command that should be run upon acceptance of this item.
 	 */
 	command?: Command;
+	/**
+	 * Identifier for the extension that contributed this completion.
+	 */
+	extensionId?: string;
```
[the remark this answers is not in the document]
Comment (digitarald) [reply src/vs/editor/common/languages.ts]: That did the trick

**Pull request merged**